### ISSUE TITLE
SSL handshake doesn't respect timeouts

### ISSUE THREAD

**Issue by feliperuiz** [lettuce-io/lettuce-core#1326]
#### Current Behavior

If SSL handshake fails, `RedisClient::connect` takes 10 seconds to fail, no matter what value I set for the timeouts.

<details>
<summary>Stack trace</summary>

```java
Exception in thread "main" io.lettuce.core.RedisConnectionException: Unable to connect to localhost:8999
        at io.lettuce.core.RedisConnectionException.create(RedisConnectionException.java:78)
        at io.lettuce.core.RedisConnectionException.create(RedisConnectionException.java:56)
        at io.lettuce.core.AbstractRedisClient.getConnection(AbstractRedisClient.java:234)
        at io.lettuce.core.RedisClient.connect(RedisClient.java:207)
        at io.lettuce.core.RedisClient.connect(RedisClient.java:192)
        at com.example.Application.main(Application.java:33)
Caused by: io.netty.handler.ssl.SslHandshakeTimeoutException: handshake timed out after 10000ms
        at io.netty.handler.ssl.SslHandler$5.run(SslHandler.java:2054)
        at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
        at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

</details>

#### Input Code

<details>
<summary>Input Code</summary>

```java
package com.example;

import io.lettuce.core.ClientOptions;
import io.lettuce.core.RedisClient;
import io.lettuce.core.RedisConnectionException;
import io.lettuce.core.RedisURI;
import io.lettuce.core.SocketOptions;
import java.time.Duration;

public class Application {
    public static void main(String[] args) {
        SocketOptions socketOptions = SocketOptions.builder()
            .connectTimeout(Duration.ofSeconds(2))
            .build();

        ClientOptions clientOptions = ClientOptions.builder()
            .socketOptions(socketOptions)
            .autoReconnect(true)
            .cancelCommandsOnReconnectFailure(true)
            .disconnectedBehavior(ClientOptions.DisconnectedBehavior.REJECT_COMMANDS)
            .build();


        RedisURI uri = RedisURI.builder()
            .withHost("localhost")
            .withPort(8999)
            .withSsl(true)
            .withTimeout(Duration.ofSeconds(2))
            .build();

        RedisClient client = RedisClient.create(uri);
        client.setOptions(clientOptions);

        var connection = client.connect();
        connection.close();
    }
}
```

</details>

#### Expected behavior/code

The connection timeout of 2 seconds should be applied and the RedisConnectionException should be thrown after the timeout.

#### Environment

- Lettuce version(s): 5.3.0.RELEASE
- Redis version: 3.2

#### Additional context

I couldn't seem to find documentation for this behavior nowhere in the docs. When searching for similar issues, I came across [this ancient issue from when SSL support was first being implemented](https://github.com/lettuce-io/lettuce-core/issues/28#issuecomment-72555736) that says:

> Build support done, connect waits for successful handshake.

Which indeed it does, but not respecting timeouts.

I know the connection timeout is part of the `SocketOptions`, which kind of implies it is only for establishing the socket connection, not what happens after it. However, if `RedisClient::connect` is only successful after a successful handshake, one could reasonably expect it to be constrained by a given connection timeout, or at least have a way of setting a timeout for the handshake.

EDIT: When creating the minimal code to reproduce the issue, I had forgotten to paste in the line that effectively sets the client options.

**Comment by mp911de** [comment]
Thanks for reporting the issue. Lettuce uses netty's default handshake timeout of 10 seconds. Applying the `RedisURI` timeout is misleading as it's not intended for connect timeouts, rather for command timeouts.

We have already `SocketOptions.getConnectTimeout()` so it would make sense to add `SslOptions.getHandshakeTimeout()` and apply it to `SslHandler`.

Do you want to submit a pull request?

**Comment by feliperuiz** [comment]
Hey, @mp911de, thanks for the reply! I can definitely submit a pull request, but I must say I'm not very familiar with the entire code, the tests and contribution protocols.

Would you be able to give me some pointers on that?

**Comment by mp911de** [comment]
Happy to review a pull request. Check out https://github.com/lettuce-io/lettuce-core/blob/main/.github/CONTRIBUTING.md for some guidance. There isn't much to consider besides tests and formatting.

**Comment by feliperuiz** [comment]
I can't seem to get all the tests to pass, even without changing anything I get 15 failures. They seem to be split into a handful of groups:

#### java.lang.reflect.UndeclaredThrowableException
<details>
Failing tests:
* io.lettuce.core.ClientOptionsIntegrationTests.authenticationTimeout


Error:
```java
java.lang.reflect.UndeclaredThrowableException
    at io.lettuce.core.ClientOptionsIntegrationTests.authenticationTimeout(ClientOptionsIntegrationTests.java:345)
Caused by: java.lang.AssertionError:

Expecting a throwable with root cause being an instance of:
 <io.lettuce.core.RedisCommandTimeoutException>
but was an instance of:
 <io.lettuce.core.RedisConnectionException: Connection closed prematurely>
    at io.lettuce.core.ClientOptionsIntegrationTests.lambda$authenticationTimeout$14(ClientOptionsIntegrationTests.java:357)
    at io.lettuce.core.ClientOptionsIntegrationTests.authenticationTimeout(ClientOptionsIntegrationTests.java:345)
```
</details>


#### java.lang.AssertionError
<details>
Failing tests:
* io.lettuce.core.pubsub.PubSubCommandTest.pingNotAllowedInSubscriptionState

Error:
```java
java.lang.AssertionError:

Expecting code to raise a throwable.
    at io.lettuce.core.pubsub.PubSubCommandTest.pingNotAllowedInSubscriptionState(PubSubCommandTest.java:496)
```
</details>


#### io.lettuce.core.RedisCommandExecutionException
<details>
Failing tests:
* io.lettuce.core.cluster.commands.ListClusterCommandIntegrationTests.lpos
* io.lettuce.core.cluster.commands.reactive.ListClusterReactiveCommandIntegrationTests.lpos
* io.lettuce.core.commands.ListCommandIntegrationTests.lpos
* io.lettuce.core.commands.reactive.ListReactiveCommandIntegrationTests.lpos
* io.lettuce.core.commands.transactional.ListTxCommandIntegrationTests.lpos

Error: 
```java
io.lettuce.core.RedisCommandExecutionException: ERR syntax error
```
</details>

#### java.net.BindException
<details>
Failing tests:
* io.lettuce.core.protocol.ConnectionFailureIntegrationTests.invalidFirstByte

Error:
```java
java.net.BindException: Address already in use
```
</details>

#### java.security.InvalidAlgorithmParameterException
<details>
Failing tests:
* io.lettuce.core.sentinel.SentinelSslIntegrationTests.shouldConnectToMasterUsingSentinel
* io.lettuce.core.SslIntegrationTests.standaloneWithJdkSslUsingTruststoreUrl
* io.lettuce.core.SslIntegrationTests.standaloneWithClientCertificates
* io.lettuce.core.SslIntegrationTests.standaloneWithJdkSsl

Error:
```java
io.lettuce.core.RedisConnectionException: Unable to connect to localhost:<PORT NUMBER>
﻿    at ...
Caused by: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
Caused by: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
```
</details>

#### io.lettuce.core.RedisConnectionException
<details>
Failing tests:
* io.lettuce.core.SslIntegrationTests.masterSlaveSslWithOneInvalidHostWillSucceed
* io.lettuce.core.SslIntegrationTests.masterSlaveWithJdkSslUsingTruststoreUrl
* io.lettuce.core.SslIntegrationTests.masterSlaveWithJdkSsl

Error: 
```java
io.lettuce.core.RedisConnectionException: Unable to connect to [...]
    at ...
Caused by: io.lettuce.core.RedisConnectionException: Failed to connect to at least one node in [...]
```
</details>

#### java.util.concurrent.TimeoutException
<details>
Failing files:
* io.lettuce.core.SslIntegrationTests.pubSubSsl

Error:
```java
java.lang.IllegalStateException: java.util.concurrent.TimeoutException: Objects are not equal: 2 and 3
    at io.lettuce.core.SslIntegrationTests.pubSubSsl(SslIntegrationTests.java:366)
Caused by: java.util.concurrent.TimeoutException: Objects are not equal: 2 and 3
    at io.lettuce.core.SslIntegrationTests.pubSubSsl(SslIntegrationTests.java:366)
```
</details>

I forked the repo, cloned the fork and followed the instructions in the **Building** section of the README. I would guess the syntax errors are due to the LPOS command being rather new, but I'm stumped about the rest. Maybe I'm missing something on my local environment? Do you think you could help me out with that?

**Comment by mp911de** [comment]
Likely some tests are not guarded properly against using older Redis versions. Feel free to submit the PR even if some of the tests fail.

**Comment by mp911de** [comment]
Merged via #1338.

**Comment by kanoonsantikul** [comment]
Hello, Is there a way to set the handshake timeout from spring properties? It seem to have the same problem when using SSL it ignore timeout properties

```
  redis:
    host: '...'
    port: 6379
    ssl: true
    timeout: 1000
```
